File: merge-tonic-lib/tonic/weave_git/hashutil.py
from __future__ import annotations

import hashlib


def normalize_lf(text: str) -> str:
    return text.replace("\r\n", "\n").replace("\r", "\n")


def canonical_text_lines(text: str) -> list[str]:
    """LF-normalized text split into lines; matches verify_path_local / manifest hashing."""
    norm = normalize_lf(text)
    lines = norm.split("\n") if norm else []
    if lines and lines[-1] == "":
        lines = lines[:-1]
    return lines


def sha256_hex_utf8(text: str) -> str:
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def sha256_hex_bytes(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()

File: packages/tonic-core/src/weaveGit/hashutil.ts
import { createHash } from "node:crypto";

export function normalizeLf(text: string): string {
  return text.replace(/\r\n/g, "\n").replace(/\r/g, "\n");
}

/** Line split after LF normalization; trailing empty line from final `\n` is dropped (parity with Python `canonical_text_lines`). */
export function canonicalTextLines(text: string): string[] {
  const norm = normalizeLf(text);
  const lines = norm.length ? norm.split("\n") : [];
  if (lines.length && lines[lines.length - 1] === "") {
    lines.pop();
  }
  return lines;
}

export function sha256HexUtf8(text: string): string {
  return createHash("sha256").update(text, "utf8").digest("hex");
}

export function sha256HexBytes(data: Buffer): string {
  return createHash("sha256").update(data).digest("hex");
}
